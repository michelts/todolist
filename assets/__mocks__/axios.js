export default {
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
};
