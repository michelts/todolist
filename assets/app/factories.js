import Factory from 'rosie';

const UserFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('username', (index) => `username_${index}`)
  .sequence('name', (index) => `Name-${index}`);

export {
  UserFactory,
};
