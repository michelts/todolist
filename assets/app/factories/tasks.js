import { Factory } from 'rosie';

const TaskFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('description', (index) => `Task ${index}`)
  .attr('priority', 1)
  .attr('due_date', '2020-05-22')
  .attr('is_completed', false);

export {
  TaskFactory,
}
