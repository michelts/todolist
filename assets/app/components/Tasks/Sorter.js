import React from 'react';
import axios from 'axios';
import { OrderedMap } from 'immutable';
import Form from 'react-bootstrap/Form';

const Sorter = ({ setTasks, disabled }) => {
  const inputRef = React.useRef();
  const [timeoutId, setTimeoutId] = React.useState(null);

  const handleChange = React.useCallback(() => {
    const { current: { value } } = inputRef;

    if (timeoutId !== null) {
      clearTimeout(timeoutId);
    }

    setTimeoutId(
      setTimeout(() => {
        setTasks(undefined);
        axios.get(`/api/v1/tasks/?sort=${value}`).then(({ data }) => {
          setTasks(
            OrderedMap(data.map((task) => [task.id, task])),
          );
        });
      }, 1000),
    );
  }, [inputRef, setTasks]);

  return (
    <Form.Control
      as="select"
      size="sm"
      custom
      ref={inputRef}
      disabled={disabled}
      onChange={handleChange}
    >
      <option value="due_date">Sort by date</option>
      <option value="priority">Sort by priority</option>
    </Form.Control>
  );
};

export default Sorter;
