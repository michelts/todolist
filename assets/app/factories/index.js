import { UserFactory } from './user';
import { TaskFactory, BlankTaskFactory } from './tasks';

export {
  UserFactory,
  TaskFactory,
  BlankTaskFactory,
};
