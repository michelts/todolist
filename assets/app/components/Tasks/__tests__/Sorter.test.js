import React from 'react';
import axios from 'axios';
import { OrderedMap } from 'immutable';
import { shallow } from 'enzyme';
import { TaskFactory } from '../../../factories';
import Sorter from '../Sorter';

jest.useFakeTimers();

const getWrapper = (customProps = {}) => {
  const props = {
    setTasks: jest.fn(),
    disabled: false,
    ...customProps,
  };
  const wrapper = shallow(<Sorter {...props} />);
  return { wrapper, props };
};

describe('Tasks component', () => {
  afterEach(() => {
    axios.get.mockClear();
  });

  it('should render a select to sort options', () => {
    const { wrapper } = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a disabled select when told through props', () => {
    const { wrapper } = getWrapper({ disabled: true });
    expect(wrapper.prop('disabled')).toBe(true);
  });

  it('should get sorted tasks after a timeout', async () => {
    // simulate a value selected in the select, once react-boostrap's onChange
    // doesn't bring the value applied
    jest.spyOn(React, 'useRef').mockReturnValue({ current: { value: 'due_date' } });

    const tasks = TaskFactory.buildList(1);
    axios.get.mockResolvedValue({ data: tasks });

    const { wrapper, props: { setTasks } } = getWrapper();
    wrapper.simulate('change');

    jest.runAllTimers();
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith('/api/v1/tasks/?sort=due_date');

    await axios.get;
    expect(setTasks).toHaveBeenCalledWith(OrderedMap(tasks.map(obj => [obj.id, obj])));
  });

  it('should reset a previous sort and get sorted tasks after a timeout', async () => {
    // simulate 2 change events by mocking 2 subsequent calls
    jest.spyOn(React, 'useRef')
      .mockReturnValueOnce({ current: { value: 'due_date' } })
      .mockReturnValueOnce({ current: { value: 'priority' } });

    const tasks = TaskFactory.buildList(1);
    axios.get.mockResolvedValue({ data: tasks });

    const { wrapper, props: { setTasks } } = getWrapper();
    wrapper.simulate('change'); // first, change to due_date
    jest.advanceTimersByTime(100);

    wrapper.simulate('change'); // second change, to priority
    jest.runAllTimers();

    // Change to due_date was ignored and his timer was resetted
    expect(axios.get).toHaveBeenCalledTimes(1);
    expect(axios.get).toHaveBeenCalledWith('/api/v1/tasks/?sort=priority');

    await axios.get;
    expect(setTasks).toHaveBeenCalledWith(OrderedMap(tasks.map(obj => [obj.id, obj])));
  });
});
