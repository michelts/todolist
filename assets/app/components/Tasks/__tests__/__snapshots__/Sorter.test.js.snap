// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks component should render a select to sort options 1`] = `
<FormControl
  as="select"
  custom={true}
  disabled={false}
  onChange={[Function]}
  size="sm"
>
  <option
    value="due_date"
  >
    Sort by date
  </option>
  <option
    value="priority"
  >
    Sort by priority
  </option>
</FormControl>
`;
